File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.py
from typing import List
from math import lcm

class Solution:
    def findKthSmallest(self, coins: List[int], k: int) -> int:

        n = len(coins)

        def count(x):
            result = 0

            for mask in range(1, 1 << n):
                value = 1
                bits = 0

                for i in range(n):
                    if mask & (1 << i):
                        bits += 1
                        value = lcm(value, coins[i])

                        if value > x:
                            break
                else:
                    if bits % 2:
                        result += x // value
                    else:
                        result -= x // value

            return result

        low = 1
        high = min(coins) * k

        while low < high:
            mid = (low + high) // 2

            if count(mid) >= k:
                high = mid
            else:
                low = mid + 1

        return low

File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.c
#include <stdlib.h>

long long gcd(long long a, long long b) {
    while (b != 0) {
        long long t = a % b;
        a = b;
        b = t;
    }

    return a;
}

long long lcm(long long a, long long b) {
    return a / gcd(a, b) * b;
}

long long countValid(
    long long x,
    int* coins,
    int n
) {
    long long result = 0;

    for (int mask = 1; mask < (1 << n); mask++) {
        long long mult = 1;
        int bits = 0;
        int valid = 1;

        for (int i = 0; i < n; i++) {
            if (mask & (1 << i)) {
                bits++;

                mult = lcm(mult, coins[i]);

                if (mult > x) {
                    valid = 0;
                    break;
                }
            }
        }

        if (!valid)
            continue;

        long long ways = x / mult;

        if (bits % 2 == 1)
            result += ways;
        else
            result -= ways;
    }

    return result;
}

long long findKthSmallest(
    int* coins,
    int coinsSize,
    int k
) {
    int minCoin = coins[0];

    for (int i = 1; i < coinsSize; i++) {
        if (coins[i] < minCoin)
            minCoin = coins[i];
    }

    long long low = 1;
    long long high = (long long)minCoin * k;

    while (low < high) {
        long long mid = low + (high - low) / 2;

        if (countValid(mid, coins, coinsSize) >= k)
            high = mid;
        else
            low = mid + 1;
    }

    return low;
}